### PR TITLE
[redo] remap non-element entities in the transformer in order to handle entities referencing each other out of order

### PR DESCRIPTION
**Pull request by MichaelBelousov**
retry of #4137

This is intended to be backported as a patch to 3.4

## Differences

- native code was updated/fixed, there is a [new native PR](https://dev.azure.com/bentleycs/iModelTechnologies/_git/imodel02/pullrequest/283251). This caused most of the (mac only) test failures.
- reuse the db or use different names for some dbs in the new ECReferenceTypesCache tests.
- some doc fixes

## What & Why

Before this PR, the transformer when copying an aspect's props will put the invalid id `0` in a navigation property if that aspect entity hasn't been mapped to the target yet. This causes crashes when processing `ExternalSourceAspects` in some models because their `scope` is expected to be valid. Worse, it has caused data loss in the past. We have a way of handling that for elements, which is putting it in the `partiallyCommittedElements` and `pendingReferences` collections and resolving those references, updating the corresponding element with the mapped id once we have it. This PR generalizes `partiallyCommittedElements` to all entities and provides some ways to refer to, insert, and update entities generically.

## Changes
- depends on a [~native PR~](https://dev.azure.com/bentleycs/iModelTechnologies/_git/imodel02/pullrequest/276050) (this PR was bad, above is the correct one)
- change `deepEqualWithFpTolerance` to `advanceEqual` with the ability to compare denormalized class names in ec props correctly, fix up small parts of older test code to use this where it makes the intent more readable
- add the `EntityReference` type, which is an Id64 but with the a root bis class tag (`ConcreteEntityTypes`) prepended, to make it possible to use them in sets of potentially overlapping id spaces (elements and non-element entities have overlapping Id64 spaces)
- `insertAspect` and corresponding aspect import functions in the transformer now returns the id (this necessitated the native change)
- introduce the future of `getReferenceIds` as the temporary `getReferenceConcreteIds` which gets an `EntityReference`. In a future breaking change, `@internal` `getReferenceConcreteIds` will be removed but its implementation will replace that of the `@beta` `getReferenceIds`.
- add an `ECReferenceTypesCache` to store the `ConcreteEntityTypes` of relationships and navigation properties, for fast retrieval where necessary, such as in the transformer
- add a small utility `makeCategorizedLogger` to make categorized logging less verbose
- add a `TupleKeyedMap` type for hash-less value-compare mapping of tuple types, which is used in `ECReferenceTypesCache`
- add a skipped test that demonstrates an in-code approach to profiling a block of code, and a utility `runWithCpuProfiler(func)`, which was used to analyze performance of changes
- minor clean up to some transformer code
- correct an out of date argument in the imodel-transformer test-app README
- add an `EntityUnifier` class to make it easier to perform CRUD (like `insert`) operations on non-abstract entities without knowing their type, as necessary for finalizing partially committed entities in the transformer
- move cloning aspects out of the transformer into the `IModelCloneContext` to make dealing with aspect entities consistent with element entities
- deprecate calling `initFromExternalSourceAspects` before any of the `process*` functions, by doing it for you
- save some milliseconds per test in a few tests by not calling `ExtensiveTestScenario.populateDb` in those tests, instead do it once lazily in `ReusedSnapshots`, and create new snapshots from it instead of creating an empty db and reinserting everything
- add a test for out of order aspect references which is what prompted this design change to handle non-element entity references. Although it doesn't happen yet afaik, this will now properly also handle navprops on models and link table relationships (although there is a bug in navprops on link table relationships as of writing)
- some changes here race https://github.com/iTwin/itwinjs-core/pull/4082, and I will have to move some of the utilities out of the backend and into the transformer to use the ecschema-metadata classes

changes since merging with master:
- add a dev dependency on ecschmea-metadata in core-backend to make a type-import to deal with raw metadata in ClassRegistry
- move ECReferenceTypesCache and EntityUnifier to transformer where I can use the ecschema-metadata package
- split deprecate `IModelCloneContext`, rename it to `IModelElementCloneContext`, and create a new `IModelCloneContext` in the transformer deriving from it that handles non-element entities.
- fix a type error in `ElementDrivesElements.create` where it used the wrong props type. Not sure why the tests didn't fail before aspect changes.
- remove detecting of relationship end reference types, since only 3 base linktable relationship classes can exist, and their constraint classes are known. I added `ModelSelectorRefersToModels` for completeness, but it isn't used and the class is sealed so can't be extended

### CODE REVIEW

**Comment by MichaelBelousov** [comment]
this is currently being held up by me not having a working mac environment to test on (my VM just crashed installing xcode)

**Comment by MichaelBelousov** [comment]
thanks to @aruniverse got it working on ARM mac. Now just need to confirm it works on x86 since as @khanaffan noticed with me, the error appears to be different there. Hopefully it's the same root cause so it's fixed.

**Comment by MichaelBelousov** [comment]
CI confirms it now works on MacOS x64

File: common/config/azure-pipelines/templates/core-build.yaml [hunk 1]
```diff
@@ -25,6 +25,71 @@ steps:
       scriptPath: common/scripts/run_rush_change.py
       arguments: $(Build.SourceBranch) $(Build.Reason) $(System.PullRequest.TargetBranch)
       failOnStderr: true
+
+  #### TEMP DO NOT MERGE
+
+  # DEPS
+
+  - script: sudo apt-get install unzip
```
Comment (MichaelBelousov) [inline common/config/azure-pipelines/templates/core-build.yaml:33]: do not resolve this comment until this has been removed. This is how we use the custom native addon.

Once this PR and the corresponding native PR is approved I will merge the native PR, queue a new native addon, remove this section, make sure CI passes under the new addon, and then merge this.

Please comment otherwise. Also @DanRod1999 please comment if I can wait for the new imodel02 workflow instead.

also note that this section would be much smaller, but nick's template is awaiting some changes to make it more cross platform, so I inlined it here with those changes.

Comment (MichaelBelousov) [reply common/config/azure-pipelines/templates/core-build.yaml:33]: Also note that the extract-api pipeline will fail until the corresponding new addon is released. iirc it's a github action and so I can't fix it the way I did the main CI jobs

**Comment by pmconne** [comment]
Waiting for builds to succeed before reviewing.

**Comment by MichaelBelousov** [comment]
> Waiting for builds to succeed before reviewing.

looks like I still had to kick them off manually even after undrafting, maybe because of the failing extract-api step...

File: common/config/azure-pipelines/templates/core-build.yaml (hunk 1 above)
Comment (DanRod1999) [reply common/config/azure-pipelines/templates/core-build.yaml:33]: unzip should now be installed in the imodeljs Linux pool. Don't wait for the new workflow, still figuring out what that will look like. I need to discuss with @calebmshafer.

**Comment by MichaelBelousov** [comment]
MacOS_Node_16 failed at step `Build iOS display-test-app swift` with `no space left on device`

**Comment by aruniverse** [comment]
> MacOS_Node_16 failed at step `Build iOS display-test-app swift` with `no space left on device`

@iTwin/mobilesdkadmins

**Comment by MichaelBelousov** [comment]
same test failed again with same message

> xcodebuild: error: Could not resolve package dependencies:
>  artifact of binary target 'IModelJsNative' failed download: The operation couldn’t be completed. No space left on device

File: common/config/azure-pipelines/templates/core-build.yaml (hunk 1 above)
Comment (MichaelBelousov) [reply common/config/azure-pipelines/templates/core-build.yaml:33]: I thought nick's PR (linked in my first comment) already makes it cross platform?

**Comment by pmconne** [comment]
@MichaelBelousov are you trying to merge your native PR, publish new addon, integrate that into master, then merge this PR into master?

**Comment by MichaelBelousov** [comment]
> @MichaelBelousov are you trying to merge your native PR, publish new addon, integrate that into master, then merge this PR into master?

Yes that is the intent I attempted to describe in the unresolved comment I posted earlier.

**Comment by pmconne** [comment]
> Yes that is the intent I attempted to describe in the unresolved comment I posted earlier.

I thought so, but that seems like a very bad idea. What are other developers supposed to do while the imodel02 branch is broken?

**Comment by MichaelBelousov** [comment]
> I thought so, but that seems like a very bad idea. What are other developers supposed to do while the imodel02 branch is broken?

I see what you mean. I will rebase this on imodel02, the primary reason to do it this way was so that I could make sure the CI passes on all platforms in this PR before it merged.

File: common/config/azure-pipelines/templates/core-build.yaml (hunk 1 above)
Comment (MichaelBelousov) [reply common/config/azure-pipelines/templates/core-build.yaml:33]: nevermind this, doing the old style imodel02 workflow, see other comments

**Comment by MichaelBelousov** [comment]
I cherry-picked only the "don't reuse empty bim changes" onto the original branch, and redid extract-api and rush lint. All other fixes are in the native branch referenced inthe description.

**Comment by MichaelBelousov** [comment]
should now be ready for review since CI doesn't apply anymore.

File: common/api/core-backend.api.md [hunk 2]
```diff
@@ -2128,6 +2174,8 @@ export class ExternalSource extends InformationReferenceElement {
     constructor(props: ExternalSourceProps, iModel: IModelDb);
     // @internal (undocumented)
     static get className(): string;
+    // (undocumented)
+    protected collectReferenceConcreteIds(referenceIds: EntityReferenceSet): void;
```
Comment (pmconne) [inline common/api/core-backend.api.md:2178]: beta function uses alpha type in signature. please find and fix all occurrences.

File: common/api/core-backend.api.md [hunk 3]
```diff
@@ -2409,7 +2461,7 @@ export abstract class GeometricElement2d extends GeometricElement {
     // @internal (undocumented)
     static get className(): string;
     // @internal (undocumented)
-    protected collectReferenceIds(referenceIds: Id64Set): void;
+    protected collectReferenceConcreteIds(referenceIds: EntityReferenceSet): void;
```
Comment (pmconne) [inline common/api/core-backend.api.md:2464]: `public` func uses `alpha` type please find and fix all.

File: common/api/core-backend.api.md [hunk 4]
```diff
@@ -2659,35 +2711,11 @@ export class HubMock {
...
-    remapCodeSpec(sourceCodeSpecName: string, targetCodeSpecName: string): void;
-    remapElement(sourceId: Id64String, targetId: Id64String): void;
-    remapElementClass(sourceClassFullName: string, targetClassFullName: string): void;
-    removeElement(sourceId: Id64String): void;
-    // @internal
-    saveStateToDb(db: SQLiteDb): void;
-    readonly sourceDb: IModelDb;
-    readonly targetDb: IModelDb;
-}
+// @beta @deprecated (undocumented)
```
Comment (pmconne) [inline common/api/core-backend.api.md:2714]: Please add to all `@deprecated` tags instructions for what to use instead.

File: common/api/core-backend.api.md (hunk 2 above)
Comment (MichaelBelousov) [reply common/api/core-backend.api.md:2178]: `collectReferenceConcreteIds` is internal, maybe I screwed up extract-api? rerunning

File: common/api/core-backend.api.md (hunk 4 above)
Comment (pmconne) [reply common/api/core-backend.api.md:2714]: nvm api-extractor calls it undocumented but the deprecated tag has this info.

File: common/api/core-backend.api.md [hunk 5]
```diff
@@ -3748,6 +3808,8 @@ export class Model extends Entity {
     constructor(props: ModelProps, iModel: IModelDb);
     // @internal (undocumented)
     static get className(): string;
+    // (undocumented)
```
Comment (pmconne) [inline common/api/core-backend.api.md:3811]: public v alpha

File: common/api/core-backend.api.md (hunk 2 above)
Comment (MichaelBelousov) [reply common/api/core-backend.api.md:2178]: looks like I need to mark every override of it internal

**Comment by toddsouthenbentley** [comment]
> same test failed again with same message
> 
> > xcodebuild: error: Could not resolve package dependencies:
> > artifact of binary target 'IModelJsNative' failed download: The operation couldn’t be completed. No space left on device

The internal disk drive for the CI machine was indeed full, Chuck cleaned it out. I think he's the only one that can do that.

File: common/api/core-backend.api.md (hunk 5 above)
Comment (MichaelBelousov) [reply common/api/core-backend.api.md]: fixed

File: common/api/core-backend.api.md (hunk 2 above)
Comment (MichaelBelousov) [reply common/api/core-backend.api.md:2178]: done

File: core/backend/src/EntityReferences.ts [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+
+  /** create the invalid id for a concrete entity type
+   * @internal
+   */
+  export function makeInvalid(type: ConcreteEntityTypes): EntityReference {
+    return `${type}${Id64.invalid}`;
+  }
+
+  // necessary to prevent cyclic dependencies, the required modules will be in the require cache already so I don't store
+  /* eslint-disable @typescript-eslint/naming-convention,@typescript-eslint/no-var-requires */
```
Comment (pmconne) [inline core/backend/src/EntityReferences.ts:72]: There's no less horrible way to solve this?

File: core/backend/src/IModelDb.ts [hunk 7]
```diff
@@ -1905,9 +1905,9 @@ export namespace IModelDb { // eslint-disable-line no-redeclare
      * @param aspectProps The properties of the new ElementAspect.
```
Comment (pmconne) [inline core/backend/src/IModelDb.ts:1905]: Document the new return type please.

Comment (MichaelBelousov) [reply core/backend/src/IModelDb.ts:1905]: I assume you mean in NextVersion.md, will do

Comment (pmconne) [reply core/backend/src/IModelDb.ts:1905]: Maybe, but I mean right here so readers of the API doc know that it `@returns the Id of the newly-inserted aspect` or whatever.

File: core/backend/src/EntityReferences.ts (hunk 6 above)
Comment (MichaelBelousov) [reply core/backend/src/EntityReferences.ts:72]: I will experiment with moving this utility to a separate file. Only `ClassRegistry` uses it, but perhaps if I separate this to another file and import in a certain order I can remove the cycles.

**Review by pmconne — APPROVED** [review]
Didn't look in depth at core-transformer changes.

File: core/backend/src/EntityReferences.ts (hunk 6 above)
Comment (MichaelBelousov) [reply core/backend/src/EntityReferences.ts]: didn't work, the cycle is in `ClassRegistry<-Entity<-EntityReference<-Model<-Element<-Entity`
instead of the getters and dynamic requires in this file, I'm going to do just one dynamic require in  `Entity.collectReferenceIds`. Please comment on that solution here and resolve if you think it is better, I just pushed it.

File: core/backend/src/IModelDb.ts (hunk 7 above)
Comment (MichaelBelousov) [reply core/backend/src/IModelDb.ts:1905]: done

File: core/backend/src/Entity.ts [hunk 8]
```diff
@@ -100,6 +100,61 @@ export class Entity {
...
+    this.collectReferenceConcreteIds(referenceIds);
+    return referenceIds;
+  }
+
+  /** Collect the Ids of this entity's *references* at this level of the class hierarchy.
+   * A *reference* is any entity referenced by this entity's EC Data
+   * This is important for cloning operations but can be useful in other situations as well.
+   * @param _referenceIds The Id64Set to populate with reference Ids.
+   * @note In the next breaking change, the behavior of this function will change to require a EntityReferenceSet argument,
+   *       which does not accept plain Id64s, see [EntityReferenceSet]($backend).
```
Comment (pmconne) [inline core/backend/src/Entity.ts:133]: `EntityReferenceSet` is defined in core-common, not core-backend; and it is `@alpha` so doc link won't be valid anyway.

Comment (MichaelBelousov) [reply core/backend/src/Entity.ts]: done

**Comment by MichaelBelousov** [comment]
I intend to perform the new addon process tomorrow morning. Can someone please comment if that means I can't merge after approvals? I believe the regular procedure is not to merge either native or js PR until you're ready to produce the addon

**Comment by aruniverse** [comment]
> > same test failed again with same message
> 
> > 
> 
> > > xcodebuild: error: Could not resolve package dependencies:
> 
> > > artifact of binary target 'IModelJsNative' failed download: The operation couldn’t be completed. No space left on device
> 
> 
> 
> The internal disk drive for the CI machine was indeed full, Chuck cleaned it out. I think he's the only one that can do that.

Why is this happening? Is the mobile sdk not getting cleaned up? Happening for other reasons unrelated to mobile build?

**Comment by toddsouthenbentley** [comment]
> > > same test failed again with same message
> > 
> > 
> > > 
> > 
> > 
> > > > xcodebuild: error: Could not resolve package dependencies:
> > 
> > 
> > > > artifact of binary target 'IModelJsNative' failed download: The operation couldn’t be completed. No space left on device
> > 
> > 
> > The internal disk drive for the CI machine was indeed full, Chuck cleaned it out. I think he's the only one that can do that.
> 
> Why is this happening? Is the mobile sdk not getting cleaned up? Happening for other reasons unrelated to mobile build?

In this case a whopping 350GB was taken up in the .npm folder. I don't know what's causing that, but it certainly isn't related to mobile builds, that was just the one that failed.

**Comment by MichaelBelousov** [comment]
@calebmshafer @paulius-valiunas could you please review this since you reviewed the previous PR https://github.com/iTwin/itwinjs-core/pull/4137?

The differences, summarized in the description, can be seen with:
- the [corresponding native PR](https://dev.azure.com/bentleycs/iModelTechnologies/_git/imodel02/pullrequest/283251) was fixed 
- sharing a filename during tests was fixed. [diff](https://github.com/iTwin/itwinjs-core/pull/4415/commits/1dc15035a1538cb3ce18e6b82614699e6f85513e)
- some doc fixes over the initial PR. [diff](https://github.com/iTwin/itwinjs-core/pull/4415/files/dcecd234da651ff990bb56e7c894edf692471d7d..7f9105779c0cbd71c2302ef7405bc66ea944db28)

Unfortunately a diff over the entire section of new commits will include the merge and therefore all changes in the imodel02 branch since the initial PR, so it is pretty big, hence the two separate diffs above.

**Review by calebmshafer — APPROVED** [review]

File: tools/eslint-plugin/dist/configs/itwinjs-recommended.js [hunk 9]
```diff
@@ -50,7 +50,14 @@ module.exports = {
     "@typescript-eslint/explicit-module-boundary-types": "off",
     "@typescript-eslint/indent": [
       "error",
-      2
+      2,
+      // SEE: https://eslint.org/docs/latest/rules/indent#options
+      {
+        // not the above documented default (0) but seems to be the default in our version of @typescript-eslint/indent
+        SwitchCase: 1,
```
Comment (calebmshafer) [inline tools/eslint-plugin/dist/configs/itwinjs-recommended.js:57]: Why did you need to change this?

Comment (MichaelBelousov) [reply tools/eslint-plugin/dist/configs/itwinjs-recommended.js:57]: specifically chained ternary expressions would keep indenting, going off memory something like this:
```ts
const w = a ? x
                   : b ? y
                        : z;
```

Comment (MichaelBelousov) [reply tools/eslint-plugin/dist/configs/itwinjs-recommended.js:57]: spoke to Caleb, removing this change atm

**Review by calebmshafer — APPROVED** [review]

File: tools/eslint-plugin/dist/configs/itwinjs-recommended.js (hunk 9 above)
Comment (MichaelBelousov) [reply tools/eslint-plugin/dist/configs/itwinjs-recommended.js]: going to do a stricter check, but `rush lint` was stalling for me (last time that happened due to having a 10GB file in my git index, which I don't this time), so I only linted common, transformer, backend, and bentley. Going to double check all affected packages.

**Comment by calebmshafer** [comment]
I overrode the policy for @ColinKerr to approve as he is out of office currently.
